CONTRIBUTING.md

```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
+ refactoring - please do these on different branches and in different PRs. While this
+ is a bit of a hassle for you, it makes your changes much easier to review and faster
+ to accept.
+ Additionally, a change history with focused commits is easier to work with in the future,
+ test failures are easier to localize and interaction between changes is easier to see.
+
+### When to Open a PR
+
+While it's OK to submit a PR directly for problems such as typos or other things where
+the motivation/problem is unambiguous, most PRs should have an associated GitHub issue.
+
+If there isn't an issue for your PR, please make an issue first and explain the problem
+ or motivation for the change you are proposing. When the solution isn't straightforward,
+ then also outline your proposed solution. Your PR will go smoother if the problem and
+ solution are agreed upon before you spend time implementing it.
+
+### Which Branch to Use
+
+Unless the issue specifically mentions a branch, please create your feature branch from `main`.
+
+For example:
+
+```sh
+# Make sure you have the most recent changes to clusterlink-net/clusterlink main
+git checkout main
+git fetch upstream main
+git rebase upstream/main
+
+# Create a branch based on main named MY_FEATURE_BRANCH
+git checkout -b MY_FEATURE_BRANCH main
+```
+
+### The Life of Pi (oops! The Life of a PR)
+
+1. Start on a [unit of work](#scope-of-a-pr) by updating from upstream and then creating
+ a branch off of it. Please use a descriptive name (e.g., `fix-issue-17` or `lb-policy-support`). Please do **not** create dependent branches and always branch from
+ an up-to-date `upstream/main`.
+
+1. You can create a draft or WIP pull request at any time. You may also
+ mark the PR as not ready by assigning the label `do-not-merge/wip` to it.
+ Reviewers will ignore it mostly unless you mention someone and ask for help.
+ Feel free to open one and use the pull request to see if the CI passes.
+ Once you are ready for a review, remove the WIP or click "Ready for Review" and
+ leave a comment that it's ready for review.
+ Give the PR a descriptive title, that would be appropriate as a commit message once
+ your work is merged. Include additional information, as appropriate, in the PR
+ description paragraph to explain what you have done.
+ If your PR fixes any issues, the description should refer to it (e.g., "fixes #NNN") or
+ you can associate it using "linked issues". Either of these links your PR to the issue
+ and automatically closes the issue when the PR is accepted.
+ Do **not** use the PR comments to document the code as they will most likely never be
+ seen after the PR is merged. Please **do** make your code self-explanatory. If there is
+ information that is needed by a programmer reading the source code, you should put it in
+ a code comment. This also applies to answering questions from reviewers: it is best done
+ by clarifying the code or add documentation, rather than in the PR comment thread.
```

Comment:
This paragraph is too long.
I would remove the last part (starting from `Do not`).

Code readability should be explained along with style, testing, etc somewhere else.